Initial support for rustc_codegen_gcc

dkm opened the pull request:
rustc can use different backend and this new script builds all the needed parts
to have GCC as the code generator.

It needs (at the moment) a patched version of GCC and builds it with libjit support.
Some glue code (in Rust) is added in the mix to finally be able to use the rustc
compiler with this dynamic backend.

Fixes https://github.com/compiler-explorer/compiler-explorer/issues/2683

dkm commented:
I think this is ready now.
It's been shellcheck'ed and tested several times using the docker image...
It simply depends on https://github.com/antoyo/rustc_codegen_gcc/pull/46 being merged by @antoyo

antoyo commented on build/build-rustc-cg-gcc.sh:
```diff
@@ -0,0 +1,189 @@
+# where the libgccjit.so will be installed
+echo "$PREFIX/lib"  > gcc_path
+
+./build_sysroot/prepare_sysroot_src.sh
```
There's now a script called `prepare_build.sh` that does this, but I guess you don't use it because of the way you install `rustup`?

dkm replied:
Yes, the script (https://github.com/antoyo/rustc_codegen_gcc/blob/master/prepare_build.sh ) adds some rustup components and calls `prepare_sysroot_src.sh`. As I also want to remove the components before packaging, I like to have the install/removal in the same script... But I can change this if you think it's better to use your script (maybe because you'll modify it?).

antoyo replied:
I don't think I'll change it, so you can keep it that way.

apmorton commented on build/build-rustc-cg-gcc.sh:
```diff
@@ -0,0 +1,189 @@
+
+RUSTUP_COMP_BUILD_DEPS=(rust-src rustc-dev llvm-tools-preview)
+
+curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- --profile minimal \
```
nitpick: can we line wrap this at the pipe so the sh continuation is not off in space?

```
curl .... |
    sh ...
```

dkm replied:
done

apmorton reviewed (changes requested):
Sorry, it didn't fully click that this is a nightly build until looking at the infra changes.

This script must implement the unchanged build skipping mechanism before I am comfortable adding it as a scheduled nightly build.

You can see this implemented here: https://github.com/compiler-explorer/misc-builder/blob/main/build/build-tinycc.sh#L36

The gist is:
- check the git revision of any cloned repos at the top of script before any other work
- build up an arbitrary string using some combination of the git revisions
- compare string to given previous build revision ($3 iirc) and bail with appropriate output if the same.

dkm commented:
> Sorry, it didn't fully click that this is a nightly build until looking at the infra changes.
> 
> This script must implement the unchanged build skipping mechanism before I am comfortable adding it as a scheduled nightly build.
> 
> You can see this implemented here: https://github.com/compiler-explorer/misc-builder/blob/main/build/build-tinycc.sh#L36
> 
> The gist is:
> 
>     * check the git revision of any cloned repos at the top of script before any other work
> 
>     * build up an arbitrary string using some combination of the git revisions
> 
>     * compare string to given previous build revision ($3 iirc) and bail with appropriate output if the same.

ha, sorry, missed that part. I should fix the mrustc builder also ! Will do that !

dkm commented:
Implemented the revision check/skip. Also refactored a bit how url/branch are used.
Currently testing it, I'll add a comment if it fails (don't see why it would...)
Will squash when review is ok :)

mattgodbolt reviewed (approved):
Looks OK to me! Thanks for taking the time!

mattgodbolt commented:
@apmorton I'm approving and you can blame me if it's not really ready :) (sorry to merge with your change request still pending, but wanted to get things moving)

dkm commented:
Thanks ! Finally got it working on my small laptop over weak connexion (took several tries), so should be ok !
